daemon/containerd/image.go

```diff
@@ -2,14 +2,223 @@ package containerd
+	"github.com/docker/docker/daemon/images"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
+	"github.com/docker/docker/layer"
+	"github.com/docker/go-connections/nat"
+	"github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
 )
 
+var truncatedID = regexp.MustCompile(`^([a-f0-9]{4,64})$`)
+
 // GetImage returns an image corresponding to the image referred to by refOrID.
-func (i *ImageService) GetImage(ctx context.Context, refOrID string, options imagetype.GetImageOpts) (retImg *image.Image, retErr error) {
-	return nil, errdefs.NotImplemented(errors.New("not implemented"))
+func (i *ImageService) GetImage(ctx context.Context, refOrID string, options imagetype.GetImageOpts) (*image.Image, error) {
+	containerdImage, img, err := i.getImage(ctx, refOrID, options.Platform)
+	if err != nil {
+		return nil, err
+	}
+
+	if options.Details {
+		size, err := containerdImage.Size(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		tagged, err := i.client.ImageService().List(ctx, "target.digest=="+containerdImage.Target().Digest.String())
+		if err != nil {
+			return nil, err
+		}
+		tags := make([]reference.Named, 0, len(tagged))
+		for _, i := range tagged {
+			name, err := reference.ParseNamed(i.Name)
+			if err != nil {
+				return nil, err
+			}
+			tags = append(tags, name)
+		}
+
+		img.Details = &image.Details{
+			References:  tags,
+			Size:        size,
+			Metadata:    nil,
+			Driver:      i.snapshotter,
+			LastUpdated: containerdImage.Metadata().UpdatedAt,
+		}
+	}
+	return img, nil
+}
+
+func (i *ImageService) getImage(ctx context.Context, refOrID string, platform *ocispec.Platform) (containerd.Image, *image.Image, error) {
+	ctrdimg, err := i.resolveImage(ctx, refOrID, platform)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	containerdImage := containerd.NewImage(i.client, ctrdimg)
+	if platform != nil {
+		containerdImage = containerd.NewImageWithPlatform(i.client, ctrdimg, platforms.OnlyStrict(*platform))
+	}
+	provider := i.client.ContentStore()
+	conf, err := ctrdimg.Config(ctx, provider, containerdImage.Platform())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	imageConfigBytes, err := content.ReadBlob(ctx, containerdImage.ContentStore(), conf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var ociimage dockerfile2llb.Image
+	if err := json.Unmarshal(imageConfigBytes, &ociimage); err != nil {
+		return nil, nil, err
+	}
+
+	fs, err := containerdImage.RootFS(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	rootfs := image.NewRootFS()
+	for _, id := range fs {
+		rootfs.Append(layer.DiffID(id))
+	}
+	exposedPorts := make(nat.PortSet, len(ociimage.Config.ExposedPorts))
+	for k, v := range ociimage.Config.ExposedPorts {
+		exposedPorts[nat.Port(k)] = v
+	}
+
+	img := image.NewImage(image.IDFromDigest(ctrdimg.Target.Digest))
```

Comment:
Just recalled https://github.com/moby/moby/pull/44426 - we should update this to

```suggestion
	img := image.NewImage(image.ID(ctrdimg.Target.Digest))
```

Reply:
Done 👍